mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,593 @@
+
+  /**
+   * All the metrics that can be currently computed by Spark for vectors.
+   *
+   * This list associates the user name, the internal (typed) name, and the list of computation
+   * metrics that need to de computed internally to get the final result.
+   */
+  private val allMetrics: Seq[(String, Metrics, DataType, Seq[ComputeMetrics])] = Seq(
+    ("mean", Mean, arrayDType, Seq(ComputeMean, ComputeWeightSum)),
+    ("variance", Variance, arrayDType, Seq(ComputeWeightSum, ComputeMean, ComputeM2n)),
+    ("count", Count, LongType, Seq()),
+    ("numNonZeros", NumNonZeros, arrayLType, Seq(ComputeNNZ)),
+    ("max", Max, arrayDType, Seq(ComputeMax, ComputeNNZ)),
+    ("min", Min, arrayDType, Seq(ComputeMin, ComputeNNZ)),
+    ("normL2", NormL2, arrayDType, Seq(ComputeM2)),
+    ("normL1", NormL1, arrayDType, Seq(ComputeL1))
+  )
+
+  /**
+   * The metrics that are currently implemented.
+   */
+  sealed trait Metrics extends Serializable
+  case object Mean extends Metrics
+  case object Variance extends Metrics
+  case object Count extends Metrics
+  case object NumNonZeros extends Metrics
+  case object Max extends Metrics
+  case object Min extends Metrics
+  case object NormL2 extends Metrics
+  case object NormL1 extends Metrics
+
+  /**
+   * The running metrics that are going to be computed.
+   *
+   * There is a bipartite graph between the metrics and the computed metrics.
+   */
+  sealed trait ComputeMetrics extends Serializable
+  case object ComputeMean extends ComputeMetrics
+  case object ComputeM2n extends ComputeMetrics
+  case object ComputeM2 extends ComputeMetrics
+  case object ComputeL1 extends ComputeMetrics
+  case object ComputeWeightSum extends ComputeMetrics
+  case object ComputeNNZ extends ComputeMetrics
+  case object ComputeMax extends ComputeMetrics
+  case object ComputeMin extends ComputeMetrics
+
+  private[stat] class SummarizerBuffer(
+      requestedMetrics: Seq[Metrics],
+      requestedCompMetrics: Seq[ComputeMetrics]
+  ) extends Serializable {
+
+    private var n = 0
+    private var currMean: Array[Double] = null
+    private var currM2n: Array[Double] = null
+    private var currM2: Array[Double] = null
+    private var currL1: Array[Double] = null
+    private var totalCnt: Long = 0
+    private var totalWeightSum: Double = 0.0
+    private var weightSquareSum: Double = 0.0
+    private var weightSum: Array[Double] = null
+    private var nnz: Array[Long] = null
+    private var currMax: Array[Double] = null
+    private var currMin: Array[Double] = null
+
+    def this() {
+      this(
+        Seq(Mean, Variance, Count, NumNonZeros, Max, Min, NormL2, NormL1),
+        Seq(ComputeMean, ComputeM2n, ComputeM2, ComputeL1,
+          ComputeWeightSum, ComputeNNZ, ComputeMax, ComputeMin)
+      )
+    }
+
+    /**
+     * Add a new sample to this summarizer, and update the statistical summary.
+     */
+    def add(instance: Vector, weight: Double): this.type = {
```

Comment:
@viirya I have tried your suggestion in the previous version code, but it do not bring performance advantage.
You can check my previous version code (in this commit "optimize summarizer buffer") and run tests.